pkg/utils/webhook/add.go

```diff
@@ -0,0 +1,158 @@
+// SPDX-FileCopyrightText: 2020 SAP SE or an SAP affiliate company and Gardener contributors.
+//
+// SPDX-License-Identifier: Apache-2.0
+
+package webhook
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/go-logr/logr"
+	admissionregistrationv1 "k8s.io/api/admissionregistration/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	ctrlwebhook "sigs.k8s.io/controller-runtime/pkg/webhook"
+)
+
+///// WEBHOOK /////
+
+// WebhookedResourceDefinition contains information about the resources that should be watched by the webhook
+type WebhookedResourceDefinition struct {
+	// APIGroup of the resource
+	APIGroup string
+	// all APIVersions of the resource that should be handled
+	APIVersions []string
+	// name of the resource, lower-case plural form
+	ResourceName string
+}
+
+// Options contains the configuration that is necessary to create a ValidatingWebhookConfiguration
+type Options struct {
+	// Name of the ValidatingWebhookConfiguration that will be created
+	WebhookConfigurationName string
+	// the webhooks will be named <resource><webhook suffix>
+	WebhookNameSuffix string
+	// base path for the webhooks, the resource name will be appended
+	WebhookBasePath string
+	// name of the service under which the webhook can be reached
+	ServiceName string
+	// namespace of the service
+	ServiceNamespace string
+	// port of the service
+	ServicePort int32
+	// LabelSelector that is used to filter all resources handled by this webhook
+	ObjectSelector metav1.LabelSelector
+	// the resources that should be handled by this webhook
+	WebhookedResources []WebhookedResourceDefinition
+	// certificates for the webhook
+	CABundle []byte
+}
+
+// ApplyValidatingWebhookConfiguration will create, update, or delete a ValidatingWebhookConfiguration, depending on the options
+// If WebhookedResources in the given options is neither nil nor empty, a ValidatingWebhookConfiguration will be created/updated
+// otherwise it will be deleted, if it exists.
+func ApplyValidatingWebhookConfiguration(ctx context.Context, mgr manager.Manager, o Options, webhookLogger logr.Logger) error {
+	tmpClient, err := getCachelessClient(mgr)
+	if err != nil {
+		return fmt.Errorf("unable to get client: %w", err)
+	}
+
+	vwc := admissionregistrationv1.ValidatingWebhookConfiguration{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: o.WebhookConfigurationName,
+		},
+	}
+
+	if o.WebhookedResources != nil && len(o.WebhookedResources) > 0 {
```

Comment:
I think we should not create any webhook resource if it is disabled so we can fail early.
```
if o.WebhookedResources == nil || len(o.WebhookedResources) == 0 {
  return nil
}
```

Reply:
OK, this basically answers my question from above. Will implement it this way.